File: ahps_web/models/module_programs.py
#
# AHPS Web - web server for managing an AtHomePowerlineServer instance
# Copyright (C) 2014, 2015  Dave Hocker (email: AtHomeX10@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the LICENSE file for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program (the LICENSE file).  If not, see <http://www.gnu.org/licenses/>.
#

from ahps_web.database.connection import get_db
from model_helpers import row_to_dict


def get_all_module_programs(houseid):
    '''
    Get all programs with relevant module information for the specified house. This is
    all of the data required to download to the AtHomePowerlineServer.
    :return: A list of module programs where each row in the list is a dict.
    '''
    db = get_db()
    sql = "select houses.houseid, modules.house_code, modules.device_code, programs.* from houses " + \
        "join rooms on rooms.houseid = houses.houseid " + \
        "join modules on modules.roomid = rooms.roomid " + \
        "join programs on programs.moduleid = modules.moduleid " + \
        "where houses.houseid=?"
    cur = db.execute(sql, [houseid])
    programs = cur.fetchall()

    program_list = []
    for row in programs:
        program_list.append(row_to_dict(row))

    return program_list

File: ahps_web/static/js/module_programs.js
/*
#
# AHPS Web - web server for managing an AtHomePowerlineServer instance
# Copyright (C) 2014, 2015  Dave Hocker (email: AtHomeX10@gmail.com)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the LICENSE file for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program (the LICENSE file).  If not, see <http://www.gnu.org/licenses/>.
#
*/

/*
    Module-programs page controller
*/
app.controller('moduleProgramsController', function($scope, $http, $sce) {
    // Initialization

    $scope.title = "AHPS Web";
    $scope.error = "";
    $scope.message = "";
    $scope.programs = [];

    // Gets the list of programs for the current module
    function get_programs() {
        var moduleid = $("#moduleid").val();
        $http.get('/module/' + String(moduleid) + '/programs', {}).
            success(function(data, status, headers, config) {
                $scope.programs = data.programs;
                // Make the HTML in the program summary safe
                for (i = 0; i < $scope.programs.length; i++) {
                    var program = $scope.programs[i];
                    program["program_summary"] = $sce.trustAsHtml(program["program_summary"]);
                }
            }).
            error(function(data, status, headers, config) {
                if (data && (data.message)) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Unable to get programs";
                }
            });
    };

    // Adds an empty program to the module. Refreshes the program
    // list to show the newly added program.
    $scope.add_program = function(moduleid) {
        $http.post('/module/' + String(moduleid) + '/program', {}).
            success(function(data, status, headers, config) {
                // Refresh the programs list
                get_programs();
            }).
            error(function(data, status, headers, config) {
                if (data && (data.message)) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Unable to get programs";
                }
            });
    };

    // Calls the edit page to edit the selected program
    $scope.edit_program = function(programid) {
        window.location.replace("/modules/program/" + String(programid) + "/page");
    };

    // Launches the dialog box to confirm remove
    $scope.remove_program = function(programid, program_name) {
        /* Set the dialog text with the room name */
        $("#dialog-text").text("Remove program: " + program_name + "?");
        /* Pop the confirmation dialog */
        $("#dialog")
            .data("programid", programid)
            .dialog("open");
    };

    // Confirmation callback that does the actual remove
    $scope.angular_remove_program = function(programid) {
        $http.delete('/module/program/' + String(programid), {}).
            success(function(data, status, headers, config) {
                // Refresh the programs list
                get_programs();
            }).
            error(function(data, status, headers, config) {
                if (data && (data.message)) {
                    $scope.error = data.message;
                }
                else {
                    $scope.error = "Unable to refresh programs";
                }
            });
    };

    // Continuing with initialization...
    get_programs();
});

/*
This code is outside of the controller because it would not
work inside the controller. Apparently there is some negative
interaction between jQuery and AngularJS.
*/
/* Initialize the confirmation dialog */
$(document).ready(function() {
    $("#dialog").dialog(
    {
        autoOpen: false,
        modal: true,
        closeOnEscape: false,
        width: 400,
        buttons: {
            "Remove": function(event) {
                $("#dialog").dialog( "close" );
                var programid = $(this).data("programid");
                // Need access to scope to call into controller
                var el = $("html");
                var scope = angular.element(el).scope();
                scope.angular_remove_program(programid);
            },
            "Cancel": function(event) {
                $("#dialog").dialog( "close" );
            }
        }
    });

});
